pages/sign_in.tsx

```diff
@@ -1,134 +1,212 @@
+      ...(await serverSideTranslations(locale, ["signInProps", "authProps"])),
     },
   };
 }
 
+//
+
 const Sign_in: NextPageWithLayout = () => {
+  const { t } = useTranslation("signInProps");
+  const { register, login } = useAuth();
+  const { getItem, setItem } = useStorage();
+  const router = useRouter();
+
+  //
+
   const [isPassprhase, setIsPassphrase] = useState(false);
+  const [isQuestions, setIsQuestions] = useState(false);
   const [step, setStep] = useState(0);
-  const [email, setEmail] = useState("");
-  const [pdfk, setPdfk] = useState("");
-  const [isMailExisting, setIsMailExising] = useState(true);
+  const [error, setError] = useState("");
 
-  const { register, login } = useAuth();
+  const [signInData, setSignInData] = useState({
+    email: "",
+    pdfk: "",
+    seed: "",
+  });
+
+  //
+
+  const emailEntered = async (data: EnterEmailNS.FormValues) => {
+    // Getting HMAC
+    const result = await register(data.email, false);
+    const key = result.keypairoomServer;
+    //
+    console.log(key);
+    if (!key) {
+      setError(t("enterEmail.notRegistered"));
+      return;
+    }
+    //
+    setSignInData({
+      ...signInData,
+      email: data.email,
+      pdfk: key,
+    });
+    //
+    setStep(1);
+  };
 
-  const errorMail = isMailExisting ? undefined : "this email doesn't exists";
   const viaPassphrase = () => {
     setIsPassphrase(true);
-    setStep(1);
+    setIsQuestions(false);
+    setStep(2);
   };
 
   const viaQuestions = () => {
     setIsPassphrase(false);
-    setStep(1);
+    setIsQuestions(true);
+    setStep(2);
   };
-  const toNextStep = async (step: number) => {
-    const result = await register(email, false);
-    if (result["keypairoomServer"]) {
-      setPdfk(result["keypairoomServer"]);
-      setStep(step);
-      devLog(result);
-    } else {
-      setIsMailExising(false);
-      devLog(result);
-    }
+
+  const passphraseEntered = (data: ViaPassphraseNS.FormValues) => {
+    setSignInData({
+      ...signInData,
+      seed: data.passphrase,
+    });
   };
 
-  const router = useRouter();
-  const { t } = useTranslation("signInProps");
+  const questionsEntered = (data: QuestionsNS.FormValues) => {
+    setSignInData({
+      ...signInData,
+      seed: getItem("seed"),
+    });
+    setStep(3);
+  };
+
+  //
+
+  // TODO: Review
+  // A fix for "passphrase" scenario - ask @bbtgnn for explanation
+  // Doing login when all the data is ready
+  useEffect(() => {
+    try {
+      // We do this only if it's passphrase mode
+      if (!isPassprhase) {
+        throw new Error("notPassphrase");
+      }
+
+      // Checking if we have all the data
+      yup
+        .object({
+          email: yup.string().required(),
+          pdfk: yup.string().required(),
+          seed: yup.string().required(),
+        })
+        .required()
+        .validateSync(signInData);
+
+      // Then logging in
+      doLogin();
+
+      //
+    } catch (error) {}
+  });
```

Comment:
Sure you don't want to have the dependencies array here? It's run just the first time after the page load?

Reply:
I would like to talk to you about this, because I'm sure there's a better way to do it. This is needed as a callback after the function `passphraseEntered` uses `setSignInData` to set the data. At first I tried putting this logic in `passphraseEntered` but somehow it didn't work.

Reply:
take look maybe at https://beta.reactjs.org/apis/react/useCallback